sdk/core/System.ClientModel/src/Convenience/ClientResultException.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.ClientModel.Primitives;
+using System.Globalization;
+using System.Runtime.Serialization;
+using System.Text;
+
+namespace System.ClientModel;
+
+[Serializable]
+public class ClientResultException : Exception, ISerializable
+{
+    private const string DefaultMessage = "Service request failed.";
+
+    private readonly PipelineResponse? _response;
+    private int _status;
+
+    /// <summary>
+    /// Gets the HTTP status code of the response. Returns. <code>0</code> if response was not received.
+    /// </summary>
+    public int Status
+    {
+        get => _status;
+        protected set => _status = value;
+    }
+
+    public ClientResultException(PipelineResponse response, Exception? innerException = default)
+        : base(GetMessage(response), innerException)
+    {
+        ClientUtilities.AssertNotNull(response, nameof(response));
+
+        _response = response;
+        _status = response.Status;
+    }
+
+    public ClientResultException(string message, PipelineResponse? response = default, Exception? innerException = default)
+        : base(message ?? DefaultMessage, innerException)
+    {
+        _response = response;
+        _status = response?.Status ?? 0;
+    }
+
+    /// <summary>
+    /// TBD
+    /// </summary>
+    /// <param name="info"></param>
+    /// <param name="context"></param>
+    protected ClientResultException(SerializationInfo info, StreamingContext context)
+        : base(info, context)
+    {
+        _status = info.GetInt32(nameof(Status));
+    }
+
+    /// <inheritdoc />
+    public override void GetObjectData(SerializationInfo info, StreamingContext context)
+    {
+        ClientUtilities.AssertNotNull(info, nameof(info));
+
+        info.AddValue(nameof(Status), Status);
+
+        base.GetObjectData(info, context);
+    }
+
+    public PipelineResponse? GetRawResponse() => _response;
+
+    private static string GetMessage(PipelineResponse response)
+    {
+        response.BufferContent();
+
+        StringBuilder messageBuilder = new();
+
+        messageBuilder
+            .AppendLine(DefaultMessage)
+            .Append("Status: ")
+            .Append(response.Status.ToString(CultureInfo.InvariantCulture));
+
+        if (!string.IsNullOrEmpty(response.ReasonPhrase))
+        {
+            messageBuilder.Append(" (")
+                .Append(response.ReasonPhrase)
+                .AppendLine(")");
+        }
+        else
+        {
+            messageBuilder.AppendLine();
+        }
+
+        // Content or headers can be obtained from raw response so are not added here.
```

Comment:
Just curious why the change from RFE, which does print the header info here.

Reply:
Yep, good call out!

The RFE exception message uses a sanitizer to prevent leaking secrets in the exception message.  We haven't added sanitization types to ClientModel yet, so I didn't want to risk leaking a secret by not having it.  Since these things are available on GetRawResponse (they weren't when we wrote the original RFE message), customers can still get those values if they needed them, so I thought this approach might be ok for a v1.  Let me know if you disagree!